### PR TITLE
Implement parallelization in register value lookup

### PR DESCRIPTION
**Pull request by koko1123**
## Goal of this PR
Implement parallelization for badger DB register lookups. I noticed the action is an atomic transaction and the integ tests cover it :)

### CODE REVIEW

File: service/index/reader.go [hunk 1]
```diff
@@ -17,6 +17,8 @@ package index
 import (
 	"errors"
 	"fmt"
+	"github.com/hashicorp/go-multierror"
```
Comment (janezpodhostnik) [inline service/index/reader.go:20]: potential linting issue

File: service/storage/operations.go [hunk 2]
```diff
@@ -190,7 +190,7 @@ func (l *Library) RetrievePayload(height uint64, path ledger.Path, payload *ledg
 		key := EncodeKey(PrefixPayload, path, height)
 		it := tx.NewIterator(badger.IteratorOptions{
 			PrefetchSize:   0,
-			PrefetchValues: true,
+			PrefetchValues: false,
```
Comment (janezpodhostnik) [inline service/storage/operations.go:193]: does this not work with prefetch?

**Review by janezpodhostnik — APPROVED** [review]

File: service/storage/operations.go (hunk 2 above)
Comment (koko1123) [reply service/storage/operations.go:193]: er eant prefetch to be a separate commit, so we know which one makes a bigger effect, before trying them together

File: service/index/reader.go [hunk 3]
```diff
@@ -98,23 +101,31 @@ func (r *Reader) Values(height uint64, paths []ledger.Path) ([]ledger.Value, err
...
-			}
-			values = append(values, payload.Value())
-		}
-		return nil
-	})
-	return values, err
+	values := make([]ledger.Value, len(paths))
+	var errs *multierror.Error
+	var wg sync.WaitGroup
+	wg.Add(len(paths))
```
Comment (SaveTheRbtz) [inline service/index/reader.go:107]: you may want to upperbound the concurrency here in cases where multiple thousands of paths are passed in

File: service/index/reader.go [hunk 4]
```diff
@@ -98,23 +101,31 @@ func (r *Reader) Values(height uint64, paths []ledger.Path) ([]ledger.Value, err
...
+					values[i] = nil
+					return nil
+				}
+				if err != nil {
+					return fmt.Errorf("could not retrieve payload (path: %x): %w", pth, err)
+				}
+				values[i] = payload.Value()
+				return nil
+			})
+			errs = multierror.Append(errs, err)
```
Comment (SaveTheRbtz) [inline service/index/reader.go:123]: You may consider preserving the fail-fast logic here.  E.g. by using https://pkg.go.dev/golang.org/x/sync/errgroup

**Review by SaveTheRbtz — APPROVED** [review]

File: service/index/reader.go [hunk 5]
```diff
@@ -98,23 +101,31 @@ func (r *Reader) Values(height uint64, paths []ledger.Path) ([]ledger.Value, err
...
+	for i, path := range paths {
+		go func(i int, pth ledger.Path) {
+			err = r.db.View(func(tx *badger.Txn) error {
+				var payload ledger.Payload
+				innerErr := r.lib.RetrievePayload(height, pth, &payload)(tx)
+				if errors.Is(innerErr, badger.ErrKeyNotFound) {
+					values[i] = nil
+					return nil
+				}
+				if err != nil {
```
Comment (SaveTheRbtz) [inline service/index/reader.go:117]:
```suggestion
				if innerErr != nil {
```

Comment (SaveTheRbtz) [reply service/index/reader.go:117]: Also you can consider using just `err` everywhere for consistency.

**Comment by koko1123** [comment]
Merging this, after some experiments I found that the path numbers rarely exceed `5000`, so I'm setting that to be the limit for active go routines @SaveTheRbtz

**Pull request merged**